### PR TITLE
backup,kvserver: skeleton stubs for testing no-copy restore

### PR DESCRIPTION
**Pull request by dt**
These changes are rough stubs, to lay out a skeleton for testing "remote" file, deferred download based RESTORE.

The changes here add a new parameter to the RESTORE statement and plumb it through to the RESTORE job, to cause it to send "remote" SSTables instead of downloading SSTables and sending their content.

The KV server is extended to handle these new remote AddSSTable requests by simply putting the location of the remote file into raft. For now, a stub implementation of handling these raft commands then goes ahead and downloads the file at the last moment before ingesting it normally.

This set of changes is primarily intended to put the rough plumbing in place, to make it easier to then iterate on the different pieces in follow-up changes, such as switching the last-second download-and-ingest out for ingesting a virtual SSTable backed by the remote file or implementing a new DistSQL processor for the restore-side sending.

Execution of this very stubbed out and unsafe implementation is gated on an env var while it contains egregious shortcuts that affect correctness (such as ignoring spans and ingesting all the keys from the whole backup, ignoring rekeys, etc).

Release note: none.
Epic: none.

### CODE REVIEW

File: pkg/ccl/backupccl/restore_job.go
Comment (itsbilal) [inline pkg/ccl/backupccl/restore_job.go:3180]: nit: `Experimetnal` here and below

File: pkg/kv/kvpb/api.proto
Comment (itsbilal) [inline pkg/kv/kvpb/api.proto:2008]: I would support `s/location/locator`, as that matches the term we're using in Pebble to find a bucket. Maybe we can also add a requirement that `location` be a path that can be parsed to create an `ExternalStorage`?

File: pkg/kv/kvpb/api.proto
Comment (itsbilal) [inline pkg/kv/kvpb/api.proto:2020]: nit: PrefixReplacement typo, here and below

**Review by itsbilal — COMMENTED** [review]
Looking good! Just some minor comments about comments

File: pkg/kv/kvpb/api.proto
Comment (msbutler) [inline pkg/kv/kvpb/api.proto:2000]: should we package these new proto fields into a single `remoteIngestion` message to prevent bloat of the addsstable request proto? I don't feel too strongly about this.

File: pkg/kv/kvserver/raftlog/payload.go
Comment (msbutler) [inline pkg/kv/kvserver/raftlog/payload.go:54]: nit: keep this check and add `&& command.ReplicatedEvalResult.RemoteFileLoc == ""`?

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go
Comment (msbutler) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go:161]: nit: return an error if args.data is not nil?

File: pkg/cloud/external_storage.go
Comment (msbutler) [inline pkg/cloud/external_storage.go:211]: if some bozo calls Open() before Init(), why block instead of error?

**Review by msbutler — COMMENTED** [review]
looks good overall! left nits and clarification questions

File: pkg/ccl/backupccl/restore_job.go
Comment (dt) [reply pkg/ccl/backupccl/restore_job.go:3183]: This isn't how long it has to return; this is how long the split point is "held" in place before it becomes eligible to be merged away if the RHS+LHS is below the target range size.

File: pkg/ccl/backupccl/restore_job.go
Comment (dt) [reply pkg/ccl/backupccl/restore_job.go:3194]: I don't have any specific vision; I think we might want to defer to kv (Erik/mira were just talking about estimated stats last week), so I don't think I have any particular insight now to record in a TODO for later.

File: pkg/cloud/external_storage.go
Comment (dt) [reply pkg/cloud/external_storage.go:211]: /shrug We need to synchronize _anyway_ so blocking is easy. AFAIK, this code is going away anyway (maybe?) as soon as pebble takes over opening/downloading the files (or maybe we should keep it and be adding a hook to pebble to allow passing in the factory implementation that it should be using to open shared file locators, so that cockroach can pass an implementation that then knows how to do nodelocal/etc files? /cc @itsbilal )

File: pkg/kv/kvpb/api.proto
Comment (msbutler) [reply pkg/kv/kvpb/api.proto:2000]: sounds good.

**Review by msbutler — APPROVED** [review]

File: pkg/cloud/external_storage.go
Comment (itsbilal) [inline pkg/cloud/external_storage.go:206]: we should move this _above_ the close right? in case the two happen concurrently. That way by the time the channel is closed, `a.factory` is set.

File: pkg/server/server.go
Comment (itsbilal) [inline pkg/server/server.go:2054]: (my earlier comment about moving up the Init is about this call)

File: pkg/cloud/external_storage.go
Comment (itsbilal) [reply pkg/cloud/external_storage.go:212]: Yeah, we create the engine before we Init the `ExternalStorageAccessor` the way this code is set up right now. So Pebble will almost certainly call `Open` before Init. I'm not the biggest fan of the synchronization either - is there any reason why we can't create the `ExternalStorageAccessor` closer to where we `CreateEngines` (and ideally before it)? That way we can just pass it into `pebble.Options.Experimental.SharedStorage` and things just fall together.

**Review by itsbilal — APPROVED** [review]
LGTM as a good starting point to get in as a way to split up future work.

**Comment by dt** [comment]
TFTRs!

bors r+

**Pull request merged**
